### PR TITLE
feat(amazonq): Add inline chat

### PR DESCRIPTION
**Pull request by grant0417**
## Problem
Adds inline chat to Amazon Q

## Solution

License: I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/amazonq/src/inlineChat/output/computeDiff.ts [hunk 1]
```diff
@@ -0,0 +1,115 @@
+/*!
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+import { type LinesOptions, diffLines, Change } from 'diff'
+import * as vscode from 'vscode'
+import { InlineTask, TextDiff } from '../controller/inlineTask'
+
+export function computeDiff(response: string, inlineTask: InlineTask, isPartialDiff: boolean): TextDiff[] | undefined {
```
Comment (justinmk3) [inline packages/amazonq/src/inlineChat/output/computeDiff.ts:9]: @ivikash is there diff-related code on master or some other branch that might overlap with this module?

File: packages/amazonq/src/inlineChat/output/responseTransformer.ts [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+
+export function responseTransformer(response: string): string | undefined {
+    try {
+        return decode(response)
+    } catch (err) {
+        if (err instanceof Error) {
+            getLogger().error(err)
+        } else {
+            getLogger().error(`An unknown error occurred: ${err}`)
+        }
```
Comment (justinmk3) [inline packages/amazonq/src/inlineChat/output/responseTransformer.ts:17]: please review https://github.com/aws/aws-toolkit-vscode/blob/master/docs/CODE_GUIDELINES.md#exceptions and https://github.com/aws/aws-toolkit-vscode/blob/master/docs/arch_develop.md#exceptions

File: packages/amazonq/src/inlineChat/provider/inlineChatProvider.ts [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+            errorMessage = AwsClientResponseError.tryExtractReasonFromSyntaxError(e) ?? defaultMessage
+        } else if (e instanceof CodeWhispererStreamingServiceException) {
+            errorMessage = e.message
+            requestID = e.$metadata.requestId
+        } else if (e instanceof Error) {
+            errorMessage = e.message
+        }
+
+        this.errorEmitter.fire()
+        void vscode.window.showErrorMessage(errorMessage)
```
Comment (justinmk3) [inline packages/amazonq/src/inlineChat/provider/inlineChatProvider.ts:155]: If code is called by a command, none of this is needed. Just throw the error, and our common handler will display it.

If throwing an error is not acceptable from some inline chat context for some reason, then at least look at the common error handling code and think about how to use it. It shouldn't be re-invented here.

File: packages/amazonq/src/inlineChat/provider/inlineChatProvider.ts [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+        getLogger().error(`error: ${errorMessage} tabID: ${tabID} requestID: ${requestID}`)
+        this.sessionStorage.deleteSession(tabID)
+    }
+
+    private async sendAuthNeededExceptionMessage(credentialState: FeatureAuthState, tabID: string, triggerID: string) {
```
Comment (justinmk3) [inline packages/amazonq/src/inlineChat/provider/inlineChatProvider.ts:160]: we have too many different ways of showing a "auth needed" message. please re-use one of the existing ones, and generalize it if needed.

Comment (grant0417) [reply packages/amazonq/src/inlineChat/provider/inlineChatProvider.ts:160]: Generalized it

File: packages/amazonq/src/inlineChat/provider/inlineChatProvider.ts (hunk 3 above)
Comment (grant0417) [reply packages/amazonq/src/inlineChat/provider/inlineChatProvider.ts:155]: Kept what logic is needed here, moved to using `ToolkitError`

**Comment by grant0417** [comment]
@justinmk3 what are next steps to merge this?

**Comment by hayemaxi** [comment]
This needs a changelog item. `npm run newChange`

**Review by zixlin7 — APPROVED** [review]

File: packages/core/src/codewhisperer/client/codewhisperer.ts [hunk 5]
```diff
@@ -226,6 +226,7 @@ export class DefaultCodeWhispererClient {
         const client = await this.createUserSdkClient()
         const requester = async (request: CodeWhispererUserClient.ListAvailableCustomizationsRequest) =>
             client.listAvailableCustomizations(request).promise()
+        // @ts-ignore
```
Comment (hayemaxi) [inline packages/core/src/codewhisperer/client/codewhisperer.ts:229]: why?

File: packages/amazonq/src/inlineChat/controller/utils.ts [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+import * as vscode from 'vscode'
+
+/**
+ * Expands the given selection to full line(s) in the document.
+ * If the selection is partial, it will be extended to include the entire line(s).
+ * @param document The current text document
+ * @param selection The current selection
+ * @returns A new Range that covers full line(s) of the selection
+ */
+export function expandSelectionToFullLines(document: vscode.TextDocument, selection: vscode.Selection): vscode.Range {
```
Comment (justinmk3) [inline packages/amazonq/src/inlineChat/controller/utils.ts:14]: this is not Q-specific. it should live in `packages/core/src/shared/utilities/textDocumentUtilities.ts`

File: packages/amazonq/src/inlineChat/controller/utils.ts [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+    const endLine = document.lineAt(selection.end.line)
+    return new vscode.Range(startLine.range.start, endLine.range.end)
+}
+
+/**
+ * Fixes the end-of-file newline for the given editor. If the selection is at the end of the
+ * last line of the document, this function Iinserts a newline character.
+ * @param editor The VS Code text editor to fix
+ */
+export async function fixEofNewline(editor: vscode.TextEditor) {
```
Comment (justinmk3) [inline packages/amazonq/src/inlineChat/controller/utils.ts:25]: docstring is unclear, and has types. also, this is not Q-specific. it should live in `packages/core/src/shared/utilities/textDocumentUtilities.ts` or `packages/core/src/shared/utilities/editorUtilities.ts`

File: packages/amazonq/src/inlineChat/output/computeDiff.ts (hunk 1 above)
Comment (ivikash) [reply packages/amazonq/src/inlineChat/output/computeDiff.ts:9]: The existing implementation generates a diff by
- Creating a temporary clone of the file in `tempDir`
- Overwriting the selected Range with the new code.

Trying to reuse the same logic, here will make it more complicated instead.

File: packages/amazonq/src/inlineChat/output/responseTransformer.ts [hunk 8]
```diff
@@ -0,0 +1,37 @@
+/*!
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import { getLogger } from 'aws-core-vscode/shared'
+import { decode } from 'he'
+import { InlineTask } from '../controller/inlineTask'
+
+export function responseTransformer(
```
Comment (justinmk3) [inline packages/amazonq/src/inlineChat/output/responseTransformer.ts:10]: no docstring for a rather non-obvious function

File: packages/amazonq/src/inlineChat/output/responseTransformer.ts [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+        } else {
+            return decodedResponse
+        }
+    } catch (err) {
+        getLogger().error('An unknown error occurred: %s', (err as Error).message)
+        return undefined
+    }
+}
+
+function extractPartialCode(response: string, inlineTask: InlineTask): [string, string] {
```
Comment (justinmk3) [inline packages/amazonq/src/inlineChat/output/responseTransformer.ts:31]: no docstring for a rather non-obvious function

File: packages/core/src/amazonq/util/authUtils.ts [hunk 10]
```diff
@@ -0,0 +1,31 @@
+/*!
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import { FeatureAuthState } from '../../codewhisperer'
+import { AuthFollowUpType, AuthMessageDataMap } from '../auth/model'
+
+export function extractAuthFollowUp(credentialState: FeatureAuthState) {
```
Comment (justinmk3) [inline packages/core/src/amazonq/util/authUtils.ts:9]: no docstring for a non-obvious function.

File: packages/core/src/shared/index.ts [hunk 11]
```diff
@@ -56,3 +56,4 @@ export { CrashMonitoring } from './crashMonitoring'
 export { amazonQDiffScheme } from './constants'
 export * from './featureConfig'
 export * from './icons'
+export * from './utilities/textDocumentUtilities'
```
Comment (justinmk3) [inline packages/core/src/shared/index.ts:59]:
```suggestion
export * as textDocumentUtil from './utilities/textDocumentUtilities'
```

**Review by justinmk3 — APPROVED** [review]

File: packages/amazonq/src/inlineChat/output/computeDiff.ts [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+        if (part.removed) {
+            if (part.value !== '\n') {
+                textDiff.push({
+                    type: 'deletion',
+                    originalText: part.value,
+                    range: new vscode.Range(startLine, 0, startLine + count, 0),
+                })
+            }
+        } else if (part.added) {
+            if (part.value !== '\n') {
```
Comment (leigaol) [inline packages/amazonq/src/inlineChat/output/computeDiff.ts:39]: can you double check if all '\n' work in Windows?

**Review by leigaol — APPROVED** [review]

File: packages/amazonq/src/inlineChat/output/computeDiff.ts (hunk 12 above)
Comment (grant0417) [reply packages/amazonq/src/inlineChat/output/computeDiff.ts:39]: This is checking the diff result which strips `\r` so this should be fine on windows.

File: packages/toolkit/package.json
[the remark this answers is not in the document]
Comment (grant0417) [reply packages/toolkit/package.json:1]: I believe this is from the icon update script, I didnt make these changes manually.

**Pull request merged**